File: 148-sort-list/sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    //important to stop at 2nd pos in case of 4 elements and 3 incase of 5 elements
    // its the reason why fast is made head.next initially
    public ListNode sortList(ListNode head) {
      if(head==null || head.next==null) return head;
      ListNode slow = head;
      ListNode fast = head.next;
      while(fast!=null && fast.next!=null)
      {
        slow = slow.next;
        fast = fast.next.next;
      }
      ListNode mid = slow.next;
      slow.next = null;
      ListNode left = sortList(mid);
      ListNode right = sortList(head);
      ListNode ans = merge(left,right);
      return ans;
    }
    public ListNode merge(ListNode left, ListNode right){
        ListNode temp = new ListNode();
        ListNode dummy = temp;
        while(left!=null && right!=null)
        {
                if(left.val<=right.val)
                {
                  dummy.next = left;
                  left = left.next;
                }
                else{
                  dummy.next = right;
                  right = right.next;
                }
                dummy = dummy.next;
        }
        if(left!=null) dummy.next = left;
        else dummy.next = right;
        return temp.next;
    }
}

File: 148-sort-list/sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    /*
     * Function: sortList
     * Purpose: Sorts a linked list using Merge Sort (Divide and Conquer).
     * Time: O(n log n)
     * Space: O(log n) for recursion stack (no extra memory for arrays).
     */
    ListNode* sortList(ListNode* head) {
        if (!head || !head->next) return head;  // Base case: empty or single node

        // Step 1: Split list into two halves
        ListNode* slow = head;
        ListNode* fast = head->next; // Important: fast starts at head->next

        /*
         * WHY fast = head->next?
         * It ensures that for even-length lists:
         * - slow stops at the 2nd node when list has 4 nodes (1,2,3,4)
         * - slow stops at the 3rd node when list has 5 nodes (1,2,3,4,5)
         * This makes splitting perfectly balanced and avoids infinite loops.
         */
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // Cut the list at mid and recurse
        ListNode* mid = slow->next;
        slow->next = nullptr;

        ListNode* left = sortList(head);   // Sort first half
        ListNode* right = sortList(mid);   // Sort second half

        // Step 2: Merge two sorted halves
        return merge(left, right);
    }

    /*
     * Function: merge
     * Purpose: Merge two sorted linked lists
     */
    ListNode* merge(ListNode* left, ListNode* right) {
        ListNode dummy(0);
        ListNode* tail = &dummy;

        while (left && right) {
            if (left->val <= right->val) {
                tail->next = left;
                left = left->next;
            } else {
                tail->next = right;
                right = right->next;
            }
            tail = tail->next;
        }

        // Append remaining nodes
        tail->next = left ? left : right;

        return dummy.next;
    }
};